tool/src/dart/class.rs

```diff
@@ -0,0 +1,914 @@
+            name,
+            body,
+            imports,
+            helper_classes,
+        }
+    }
+}
+
+pub struct TyGenContext<'a, 'dartcx, 'tcx> {
+    cx: &'dartcx DartContext<'tcx>,
+    imports: &'a mut BTreeSet<Import>,
+    helper_classes: &'a mut BTreeMap<String, String>,
+}
+
+impl<'a, 'dartcx, 'tcx: 'dartcx> TyGenContext<'a, 'dartcx, 'tcx> {
+    fn gen_enum(&mut self, ty: &'tcx hir::EnumDef, id: TypeId, type_name: &str) -> String {
+        #[derive(Template)]
+        #[template(path = "dart/enum.dart.jinja", escape = "none")]
+        struct ImplTemplate<'a> {
+            ty: &'a hir::EnumDef,
+            fmt: &'a DartFormatter<'a>,
+            type_name: &'a str,
+            methods: &'a [MethodInfo<'a>],
+            docs: String,
+        }
+
+        let methods = ty
+            .methods
+            .iter()
+            .flat_map(|method| self.gen_method_info(id, method, &type_name))
+            .collect::<Vec<_>>();
+
+        ImplTemplate {
+            ty,
+            fmt: &self.cx.formatter,
+            type_name: &type_name,
+            methods: methods.as_slice(),
+            docs: self.cx.formatter.fmt_docs(&ty.docs),
+        }
+        .render()
+        .unwrap()
+    }
+
+    fn gen_opaque_def(&mut self, ty: &'tcx hir::OpaqueDef, id: TypeId, type_name: &str) -> String {
+        #[derive(Template)]
+        #[template(path = "dart/opaque.dart.jinja", escape = "none")]
+        struct ImplTemplate<'a> {
+            type_name: &'a str,
+            methods: &'a [MethodInfo<'a>],
+            docs: String,
+            destructor: String,
+        }
+
+        self.imports.extend([Import {
+            path: "dart:ffi".into(),
+            suffix: " as ffi".into(),
+        }]);
+
+        let methods = ty
+            .methods
+            .iter()
+            .flat_map(|method| self.gen_method_info(id, method, &type_name))
+            .collect::<Vec<_>>();
+
+        let destructor = self.cx.formatter.fmt_destructor_name(id);
+
+        ImplTemplate {
+            type_name: &type_name,
+            methods: methods.as_slice(),
+            destructor,
+            docs: self.cx.formatter.fmt_docs(&ty.docs),
+        }
+        .render()
+        .unwrap()
+    }
+
+    fn gen_struct_def<P: TyPosition>(
+        &mut self,
+        ty: &'tcx hir::StructDef<P>,
+        id: TypeId,
+        type_name: &str,
+    ) -> String {
+        #[derive(Template)]
+        #[template(path = "dart/struct.dart.jinja", escape = "none")]
+        struct ImplTemplate<'a> {
+            type_name: &'a str,
+            fields: &'a [FieldInfo<'a>],
+            methods: &'a [MethodInfo<'a>],
+            docs: String,
+        }
+
+        struct FieldInfo<'a> {
+            name: Cow<'a, str>,
+            annotation: Option<Cow<'a, str>>,
+            ffi_cast_type_name: Cow<'a, str>,
+            dart_type_name: Cow<'a, str>,
+            get_expression: Cow<'a, str>,
+            set_cleanups: Vec<String>,
```

Comment:
thought: hm, i'm not sure if we should lazy-convert these structs as opposed to eagerly converting them into pure dart structs. especially for things like string conversions. thoughts?

Reply:
Hmm, it's a tradeoff because for argument structs, this *is* the lazy conversion

Reply:
wait no, other way around, for return structs this is the lazy conversion